src/ucp/wireup/wireup_cm.c

```diff
@@ -197,9 +197,60 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
     ucs_error("UCP CM wireup is not completely implemented");
 }
 
+static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
+{
+    ucp_ep_h ucp_ep = req->send.ep;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    ucp_ep_cm_disconnect(ucp_ep);
+    ucp_request_complete_send(req, UCS_OK);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+}
+
+static unsigned ucp_ep_cm_disconnect_progress(void *arg)
+{
+    ucp_ep_h ucp_ep     = (ucp_ep_h)arg;
+    ucp_worker_h worker = ucp_ep->worker;
+    void *req;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    ucs_trace_func("ucp_ep = %p, ucp_ep->flags = %xu", ucp_ep, ucp_ep->flags);
+
+    ucp_ep->flags &= ~UCP_EP_FLAG_REMOTE_CONNECTED;
+
+    if (ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
+        req = ucp_ep_flush_internal(ucp_ep, UCT_FLUSH_FLAG_LOCAL, NULL, 0, NULL,
+                                    ucp_ep_cm_disconnect_flushed_cb,
+                                    "disconnected_cb");
+        if (req == NULL) {
+            ucp_ep_cm_disconnect(ucp_ep);
+        } else if (UCS_PTR_IS_PTR(req)) {
+            ucp_request_release(req);
+        } else {
+            ucs_fatal("ucp_ep_flush_internal completed with error: %s",
+                      ucs_status_string(UCS_PTR_STATUS(req)));
+        }
+    } else {
+        ucp_ep_do_disconnect(ucp_ep_ext_gen(ucp_ep)->close_req.req);
+    }
+
+    UCS_ASYNC_UNBLOCK(&worker->async);
+
+    return 1;
+}
+
 static void ucp_cm_disconnect_cb(uct_ep_h ep, void *arg)
 {
-    ucs_error("UCP close protocol is not completely implemented");
+    ucp_ep_h ucp_ep            = (ucp_ep_h)arg;
+    uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    ucs_assert(ucp_ep_get_cm_uct_ep(ucp_ep) == ep);
+
+    /* invoke the disconnect flow from the main thread */
+    uct_worker_progress_register_safe(ucp_ep->worker->uct,
```

Comment:
we cannot poll all EPs which are in disconnect process all the time

Reply:
we don't, this progress just schedules flush then we poll flush is completed as usual.
will rename callback

Reply:
pls add more code doc in this function

Reply:
ls add more code doc in this function